crates/store/src/state.rs

```diff
@@ -96,17 +96,114 @@ pub struct TransactionInputs {
+                    highest_block_num,
+                    latest_block_num,
+                });
+            }
+
+            // Remove the latest block from the to-be-tracked blocks as it will be the reference
+            // block for the batch itself and thus added to the MMR within the batch kernel, so
+            // there is no need to prove its inclusion.
+            blocks.remove(&latest_block_num);
+
+            // Using latest block as the target forest means we take the state of the MMR one before
+            // the latest block. This is because the latest block will be used as the reference
+            // block of the batch and will be added to the MMR by the batch kernel.
+            let target_forest = latest_block_num.as_usize();
+            let peaks = self
+                .get()
+                .peaks_at(target_forest)
+                .expect("target_forest should be smaller than forest of the chain mmr");
+            let mut partial_mmr = PartialMmr::from_peaks(peaks);
+
+            for block_num in blocks.iter().map(|arg0: &BlockNumber| BlockNumber::as_usize(&arg0)) {
+                // SAFETY: We have ensured block nums are less than chain length.
+                let leaf = self
+                    .get()
+                    .get(block_num)
+                    .expect("block num less than chain length should exist in chain mmr");
+                let path = self
+                    .get()
+                    .open_at(block_num, target_forest)
+                    .expect("block num and target forest should be valid for this mmr")
+                    .merkle_path;
+                // SAFETY: We should be able to fill the partial MMR with data from the chain MMR
+                // without errors, otherwise it indicates the chain mmr is invalid.
+                partial_mmr
+                    .track(block_num, leaf, &path)
+                    .expect("filling partial mmr with data from mmr should succeed");
+            }
+
+            (latest_block_num, partial_mmr)
+        };
+
+        // TODO: Unnecessary conversion. We should change the select_block_headers function to take
+        // an impl Iterator instead to avoid this allocation.
+        let mut blocks: Vec<_> = blocks.clone().into_iter().collect();
+        // Fetch the reference block of the batch as part of this query, so we can avoid looking it
+        // up in a separate DB access.
+        blocks.push(batch_reference_block);
+        let headers = db
+            .select_block_headers(blocks)
+            .await
+            .map_err(GetBatchInputsError::SelectBlockHeaderError)?;
+
+        let chain_mmr =
+            ChainMmr::new(partial_mmr, headers.clone()).expect("Unable to build ChainMmr");
+        Ok((chain_mmr, headers, batch_reference_block))
+    }
+}
 /// Container for state that needs to be updated atomically.
 struct InnerState {
     nullifier_tree: NullifierTree,
-    chain_mmr: Mmr,
+    chain_mmr: Blockchain,
```

Comment:
Can we rename this field to `blockchain` instead? `chain_mmr` suggests it is a `ChainMmr` but it's not.